### ISSUE TITLE
ref strategy reasons.

### ISSUE THREAD

**Issue by blade2005** [ecyrbe/zodios-openapi#108]
Hi! it's me again. this time on this repo.

I see in the code base `$refStrategy` is set to an explicit "none" I was wondering what the reason was for that? I'm used to see references used as much as possible in openapi documents.

**Comment by blade2005** [comment]
If there wasn't a solid reason would it be possible to allow us to set it ourselves?

**Comment by ecyrbe** [comment]
no valid reason, it could indeed be user choice.

**Comment by blade2005** [comment]
An attempt to implement this leads me to believe a change would be needed to achieve this.

One of the core approaches would be having a reoccurring schema object(map) to pass to `zodToJsonSchema`.

Would probably need a refactor of how `makeOpenApi` works. In my mind and the strategy I'd use is to capture all schema's and responses into a definitions section(would require no duplicate names). Which means we'd need a way on the zod object to set a name to use.

I was afraid of this. I now understand why [@asteasolutions/zod-to-openapi](https://github.com/asteasolutions/zod-to-openapi) took the approach they did.

One of the things they did was 
1. add a registry
2. add a `openapi` method on `zod` objects.

allowing to register a param to be re-used and how they did direct look ups makes sense. I feel like with a custom method to store additional data would make this approach more feasible.

I think this may be a nice to have but represents a large block of work.

I'd like to keep this open as I think it would be good to have.

**Comment by blade2005** [comment]
I created this. and it is technically correct but since the converts small bits to openapi most things dont make it to references.

https://github.com/blade2005/zodios-openapi/tree/feature/108/user-settable-ref-strategy

**Comment by ecyrbe** [comment]
Thanks, that's awesome that you took the Time to look into it.
I'll try to take also a deep look into what you did when i'll have more Time (most likely this week-end)

**Comment by blade2005** [comment]
To be honest, if we had a label option on zod types this would be really easy. https://github.com/colinhacks/zod/issues/1767

```typescript
  const defs: Record<string, z.ZodType<any>> = {};
  for (let api of options.apis) {
    for (let endpoint of api.definitions) {
      defs[endpoint.response.label] = endpoint.response;
      for (let error of endpoint.errors ?? []) {
        defs[error.schema.label] = error.schema;
      }
      // do same idea for params (path, query, header)
      // add body into defs
    }
  }
  // repeat loop but this time when calling `makeJsonSchema` pass in defs
```

I'd probably abstract that set of loops into it's own function and return defs or pass it in.

but alas without label or adding in something that can give a title to the schema it makes this difficult.

**Comment by lnogol** [comment]
I think this is important since openapi client generators often make use of the schema name

> To be honest, if we had a label option on zod types this would be really easy. https://github.com/colinhacks/zod/issues/1767

Could it perhaps be implemented using a custom convention on `.describe()`? Such as

```ts
schema.describe("label=User; Description for user")
```

that way we could have both, until zod implements `.label()` (if ever)

**Comment by blade2005** [comment]
@lnogol not a bad idea.

Doesn't zodios-openapi use the description for the schema too?

**Comment by lnogol** [comment]
I believe it's used internally by `zod-to-json-schema` dep. It ends up as `description` in openapi. `zodios-openapi` would have to strip the label away before passing the schema to `zod-to-json-schema`

Not sure how this could work recursively though, such as labels on keys of an object schema

**Comment by ecyrbe** [comment]
This is not an easy topic, i'm still trying to check what is the right course of action. 
All ideas are welcomed.
Be aware that with Zodios v11 i want to make zodios agnostic to validation library. 
This will work well for everything but openapi because zod is the only library that does openapi kinda well (at least better than other libs)
But if you have any idea to make openapi generation the most independant of zod (and working with anything) i'm all ears open.

**Comment by blade2005** [comment]
an object schema should result in a ref, but the keys there of probably shouldn't. though they could in some cases it's needed/wanted. 

The code would have to recursively delve into the schema's looking for descriptions which start with `label=` and create a def for those. but you'd need to do it as depth first, then the next layer back up, this avoids issues where you might create a def for something which itself uses another def....

It start getting very convoluted.

@ecyrbe regarding openapi generation which is independent from any validation library, you got me there.... unless you build adapter classes for different validations to fit into something which can generate openapi code I don't see it happening.

**Comment by lnogol** [comment]
This turned out to be an issue for us. The Swift openapi generator's output was a bunch of `InlineObject{number}DTO.swift` models (for request and response bodies) which is not very useful. The models could also randomly renumber anytime there was a change to the openapi doc.

@blade2005 mentioned [@asteasolutions/zod-to-openapi](https://github.com/asteasolutions/zod-to-openapi) so I gave it a shot and I'm now generating openapi using that library, while still using zodios for everything else.

in api definition
```ts
import { makeApi } from "@zodios/core";
import { OpenAPIRegistry } from "@asteasolutions/zod-to-openapi";

export const registry = new OpenAPIRegistry();
export const makeSchema = <TSchema extends ZodSchema>(name: string, schema: TSchema): TSchema => {
  return registry.register(name, schema);
};

export const UserSchema = makeSchema("User", z.object(...));

export default makeApi([
  {
    method: "get",
    path: "/users",
    alias: "getUsers",
    response: z.array(UserSchema)
  },
])
```

in openapi module
```ts
import { OpenAPIGenerator } from "@asteasolutions/zod-to-openapi";
import { ZodiosEndpointDefinition } from "@zodios/core";
import { z, ZodVoid, ZodSchema } from "zod";

import api from "@/api";
import { registry } from "@/api/common/helpers";

const capitalize = (s: string) => s && s[0].toUpperCase() + s.slice(1);
const pathRegExp = /:([a-zA-Z_][a-zA-Z0-9_]*)/g;

const bearerAuth = registry.registerComponent("securitySchemes", "BearerAuth", {
  type: "http",
  scheme: "bearer",
  bearerFormat: "JWT"
});

(api as (ZodiosEndpointDefinition & { metadata?: Record<string, string> })[]).forEach(
  (endpoint) => {
    const path = endpoint.path.replace(pathRegExp, "{$1}");
    const params = endpoint.parameters?.filter((p) => p.type === "Path") || [];
    const query = endpoint.parameters?.filter((p) => p.type === "Query") || [];
    const body = endpoint.parameters?.find((p) => p.type === "Body");

    const schemaName = [
      capitalize(endpoint.method),
      ...endpoint.path
        .replace(pathRegExp, "$1")
        .split("/")
        .map((s) =>
          s
            .split("-")
            .map((a) => capitalize(a))
            .join("")
        )
    ].join("");

    // TODO:
    if (endpoint.requestFormat === "form-data") return;

    registry.registerPath({
      // @ts-ignore
      method: endpoint.method,
      path,
      tags: [path.split("/")[1]],
      // @ts-ignore
      description: endpoint.description || undefined,
      // @ts-ignore
      summary: endpoint.description || undefined,
      security: [{ [bearerAuth.name]: [] }],
      request: {
        ...(params.length > 0
          ? {
              params: z.object(
                params.reduce((acc, val) => {
                  acc[val.name] = val.schema.openapi({ param: { description: val.description } });
                  return acc;
                }, {} as Record<string, ZodSchema>)
              )
            }
          : {}),
        ...(query.length > 0
          ? {
              query: z.object(
                query.reduce((acc, val) => {
                  acc[val.name] = val.schema.openapi({ param: { description: val.description } });
                  return acc;
                }, {} as Record<string, ZodSchema>)
              )
            }
          : {}),
        ...(body
          ? {
              body: {
                description: body.description,
                content: {
                  "application/json": {
                    schema: registry.register(`${schemaName}Body`, body.schema)
                  }
                }
              }
            }
          : {})
      },
      responses: {
        [endpoint.status || 200]: {
          description: endpoint.responseDescription || "Success",
          ...(endpoint.response instanceof ZodVoid
            ? {}
            : {
                content: {
                  [endpoint.metadata?.contentType || "application/json"]: {
                    schema: registry.register(`${schemaName}Response`, endpoint.response)
                  }
                }
              })
        }
      }
    });
  }
);

const generator = new OpenAPIGenerator(registry.definitions, "3.0.0");
export const openApi = generator.generateDocument({ info: { ... }});
```

maybe not perfect, but generates usable openapi and works so far

the only problem I see with this is that the `OpenAPIRegistry` is imported on the client side

**Comment by blade2005** [comment]
@lnogol great idea, i like it.

I'm using prefixed routes.

```
router.use("/users", userRouter);
app.use(router);
```

How can I get a list of all the finalized api's to do for the `for each`?

Or would I need to call something like this `prefixApi("/users", usersApi)` to get those outside of the router/app?

**Comment by lnogol** [comment]
sorry, I'm not familiar with prefixed routes. I use full paths everywhere for simplicity

but looking at the [implementation](https://github.com/ecyrbe/zodios-express/blob/main/src/zodios.utils.ts#L24-L32) I would say yes. run all your "sub apis" through `prefixApi`, flatten the result and then call the `forEach` above

**Comment by blade2005** [comment]
> but looking at the [implementation](https://github.com/ecyrbe/zodios-express/blob/main/src/zodios.utils.ts#L24-L32) I would say yes. run all your "sub apis" through `prefixApi`, flatten the result and then call the `forEach` above

@lnogol thanks!

Also was taking a look at the code you provided.

One of the things I did in my use of `@asteasolutions/zod-to-openapi` was to register at declaration all of the schemas and defined the request and response schema's in one place next to your internal schemas. this also means if you have other endpoints which return the same schema as another you don't duplicate the schema in the components section in the openapi spec. 

```
export const UserSchema = makeSchema(
  "User",
  z.object({
    id: z.number(),
    name: z.string(),
    isAdmin: z.boolean()
  })
);

export const UserResponseSchema = makeSchema(
  "UserResponse",
  UserSchema.omit({ isAdmin: true }).strip()
);
```

I don't want to get too far off topic of the issue.

Bringing it back to zodios, I think the best approach based on how zodios-openapi is set up is to wait on this until a label method gets supported. I suspect though if zodios is moving away from being zod specific that openapi support may not be entirely feasible
